# Lab 2 - Analysis of Nuclear and Alternative Energy use as a Percentage of Total Energy Consumption by Country 
## David Abramowitz 9/20/2018

## Abstract

In this lab, I looked at data from the World Bank on countries' percentage of nuclear and alternative energy use compared to their total energy use. The average percentage and the minimum and maximum percentages were found for the years 2011 and 2012. It was found that Iceland uses the most alternative and/or nuclear energy with over 89% of its energy use being comprised of these sources. The minimum percentages were both zero and the average for both 2011 and 2012 were over eight percent. 

## Dataset Preparation

This data was accessed through the World Bank website, which can be found here: https://data.worldbank.org/indicator/EG.USE.COMM.CL.ZS?view=chart. The data itself came from the International Energy Agency and the Organization for Economic Cooperation and Development. Here is the original CSV file: API_EG.USE.COMM.CL.ZS_DS2_en_csv_v2_10082393.csv. This data represents the percentage of nuclear and alternative energy use. So, if half of a country's energy production came from a nuclear or alternative source then its percentaeg would be 50. Note that this figure does not take into consideration the volume of energy production. For example, if the country with half of its energy coming form nuclear or alternative sources is Country A and a country with a quarter of its energy coming from those sources is Country B, then Country A's percentage will be 50 to B's 25. However Country B may produce two times as much energy as A, so while B's percentage is half that of A's, the energy outpute is equal.  

Here is a link to data in Google Sheets: https://docs.google.com/spreadsheets/d/1zKy34i5Jsp_heEU76nN3KP3mTYfUguFIKJKJ671mB7k/edit?usp=sharing
The names in red are those that are suspected to be noncountries, and those in yellow are territories, protectorates, or other areas where confusion may arise. The four rows of hear data are intact as they do not get removed in Google Sheets. 

This dataset included indexes, groups of countries that created several redundancies. Becuase my focus was on indivudal countries, these noncountries had to be removed. Below are all of the names that were removed:
- Arab World
- Central Europe and the Baltics
- Caribbean small states
- East Asia & Pacific (excluding high income)
- Early-demographic dividend
- East Asia & Pacific
- Europe & Central Asia (excluding high income)
- Europe & Central Asia
- Euro area
- European Union
- Fragile and conflict affected situations
- High income
- Heavily indebted poor countries (HIPC)
- IBRD only
- IDA & IBRD total
- IDA total
- IDA blend
- IDA only
- Not classified
- Latin America & Caribbean (excluding high income)
- Latin America & Caribbean
- Least developed countries: UN classification
- Low income
- Lower middle income
- Low & middle income
- Late-demographic dividend
- Middle East & North Africa
- Middle income
- Middle East & North Africa (excluding high income)
- North America
- OECD members
- Other small states
- Pre-demographic dividend
- Pacific island small states
- Post-demographic dividend
- South Asia
- Sub-Saharan Africa (excluding high income)
- Sub-Saharan Africa
- Small states
- East Asia & Pacific (IDA & IBRD countries)
- Europe & Central Asia (IDA & IBRD countries)
- Latin America & the Caribbean (IDA & IBRD countries)
- Middle East & North Africa (IDA & IBRD countries)
- South Asia (IDA & IBRD)
- Sub-Saharan Africa (IDA & IBRD countries)
- Upper middle income
- World

Territories such as Puerto Rico were counted seperately. Below is a list of territories/protectorates and other areas of potential confusion. All of these were counted as being seperate to avoid confusion. 
- American Samoa
- Channel Islands
- Hong Kong SAR, China
- Macao SAR, China
- New Caledonia
- Puerto Rico
- West Bank and Gaza
- French Polynesia
- Sint Maarten (Dutch part)
- Turks and Caicos Islands
- British Virgin Islands
- Virgin Islands (U.S.)


In [31]:
#noncountries manually deleted in Google Sheets

In [32]:
#open the file
data_file = open("Untitled spreadsheet - API_EG.USE.COMM.CL.ZS_DS2_en_csv_v2_10082393.csv", "r")

#make an empty list to store the file
energy_list = [];

#enter all of the lines from the CSV file into the list
for row in data_file:
    energy_list.append(row);
    
#close the file
data_file.close();

#check that the file was properly read into the list
print(energy_list[:8])


['Data Source,World Development Indicators,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,\n', ',,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,\n', 'Last Updated Date,2018-08-28,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,\n', ',,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,\n', 'Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,1966,1967,1968,1969,1970,1971,1972,1973,1974,1975,1976,1977,1978,1979,1980,1981,1982,1983,1984,1985,1986,1987,1988,1989,1990,1991,1992,1993,1994,1995,1996,1997,1998,1999,2000,2001,2002,2003,2004,2005,2006,2007,2008,2009,2010,2011,2012,2013,2014,2015,2016,2017\n', 'Aruba,ABW,Alternative and nuclear energy (% of total energy use),EG.USE.COMM.CL.ZS,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,\n', 'Afghanistan,AFG,Alternative and nuclear energy (% of total energy use),EG.USE.COMM.CL.ZS,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,\n', 'Angola,AGO,Alte

In [33]:
#delete the top five rows as they are the header data using a loop
for index in range (5):
    del energy_list[0];

#check that the rows were properly deleted by printing some data
print(energy_list[:5])

['Aruba,ABW,Alternative and nuclear energy (% of total energy use),EG.USE.COMM.CL.ZS,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,\n', 'Afghanistan,AFG,Alternative and nuclear energy (% of total energy use),EG.USE.COMM.CL.ZS,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,\n', 'Angola,AGO,Alternative and nuclear energy (% of total energy use),EG.USE.COMM.CL.ZS,,,,,,,,,,,,1.351166059,1.410645793,1.694729788,1.748593374,1.982569886,0.9038063093,0.9167393091,1.028037849,1.065821929,1.121397344,1.114046147,1.197733519,1.067446156,1.098589514,1.208084947,1.194329205,1.166167561,1.120664063,1.110308595,1.059805315,1.102117333,1.192545009,1.220019931,1.210327457,1.218061875,1.216598781,1.126308332,1.367893399,1.098361843,1.090812748,1.165133129,1.251680189,1.259392698,1.726128644,2.264856497,2.479161747,2.177703582,2.525101679,2.293878407,2.61723535,2.726740489,2.339021054,2.952296051,2.954634798,,,\n', 'Albania,ALB,Alternative and nuclear energy (% of total energy use),E

After the data was downloaded and header data (top five rows) got deleted, some of data was printed to ensure that it had been properly imported. Four seperate lists were created for the data; a list for the countries, a list for 2011 energy data, a list for 2012 energy data, and a list for nonreporting countries. A list for noncountries was then created.  

In [34]:
#create seperate lists for 2011 and 2012 and the countries and nonreporters 
country_list = [];
year_2011_list = [];
year_2012_list = [];
nonreporting_countries_2011_2012_list = [];

#making a list of codes for noncountries/regions
#I don't include these entities in my statistical calculations 
noncountry_codes_list = ["ARB", "CEB", "CSS", "EAP", "EAR", "EAS", "ECA", "ECS", "EMU", "EUU", "FCS", "HIC", "HPC", "IBD", "IBT", "IDA", "IDB", "IDX", "INX", "LAC", "LCN", "LDC", "LIC", "LMC", "LMY", "LTE", "MEA", "MIC", "MNA", "NAC", "OED", "OSS", "PRE", "PSS", "PST", "SAS", "SSA", "SSF", "SST", "TEA", "TEC", "TLA", "TMN", "TSA", "TSS", "UMC", "WLD"]

#A for loop to pull the information out of the original energy list
for item in range(len(energy_list)):
    #split up the data
    energy_list[item] = energy_list[item].split(",")
    
    
   #create variables for the country names and codes
    country_name = energy_list[item][0]
    country_code = energy_list[item][1]
    #isolate 2011 and 2012 data from the noncountry data and add them into their repsective lists
    if not(country_code in noncountry_codes_list): 
   
        
        
   
        year_2011_item = energy_list[item][55]
        year_2012_item = energy_list[item][56]
     #check to make sure that the country code is not for a region or noncountry using the noncountry codes 
     #if data is blank in the 2011, 2012, or country list, it will go into the nonreporting list
     #if data is NOT blank it will go to the 2011, 2012, or country list 
        if year_2011_item == '' or year_2012_item == '':   #data is missing
            nonreporting_countries_2011_2012_list.append(country_name)  
        else:
            year_2011_list.append(float(year_2011_item))
            year_2012_list.append(float(year_2012_item))
            country_list.append(country_name)
   #check that noncountries were isloated correctly
    else:
        print("Not a country:", country_name)
    

#check that lists were split correctly
print(year_2011_list[:5])
print(year_2012_list[:5])
print(country_list[:5])
print(nonreporting_countries_2011_2012_list[:5])

Not a country: Arab World
Not a country: Central Europe and the Baltics
Not a country: Caribbean small states
Not a country: East Asia & Pacific (excluding high income)
Not a country: Early-demographic dividend
Not a country: East Asia & Pacific
Not a country: Europe & Central Asia (excluding high income)
Not a country: Europe & Central Asia
Not a country: Euro area
Not a country: European Union
Not a country: Fragile and conflict affected situations
Not a country: High income
Not a country: Heavily indebted poor countries (HIPC)
Not a country: IBRD only
Not a country: IDA & IBRD total
Not a country: IDA total
Not a country: IDA blend
Not a country: IDA only
Not a country: Not classified
Not a country: Latin America & Caribbean (excluding high income)
Not a country: Latin America & Caribbean
Not a country: Least developed countries: UN classification
Not a country: Low income
Not a country: Lower middle income
Not a country: Low & middle income
Not a country: Late-demographic dividend


The codes for noncountries were manually collected from data and entered into the noncountries list. The data was split and variables were created for the country codes. After that, data for 2011 and 2012 was isolated (this was done by manually counting which rows this data was stored in, it got checked to ensure that the rows were not miscounted). If a country's code matched a code in the noncountries list, it was not included in statistical calculations. In the 2011 and 2012 lists, nonreporting countries were frequent. As such, if an item in either list was blank it got added into the nonreporting list. Noncountries were listed to ensure accuracy. 

## Data Modeling
The maximum percenatges in the 2011 and 2012 lists were found and their corresponding country names were also found. Additionally, the minimum percentages and the associated country names were also found.  

In [35]:
#find highest percentages for 2011 and 2012 
max_energy_percentage_2011 = max(year_2011_list)
print("Max percentage for 2011 is", max_energy_percentage_2011)

max_energy_percentage_2012 = max(year_2012_list)
print("Max percentage for 2012 is", max_energy_percentage_2012)

#find the corresponding country names
max_index = year_2011_list.index(max_energy_percentage_2011)
print("Country with the maximum percentage in 2011 is", country_list[max_index])

max_index = year_2012_list.index(max_energy_percentage_2012)
print("Country with the maximum percentage in 2012 is", country_list[max_index])

#finding the minimum percentages for 2011 and 2012
min_energy_percentage_2011 = min(year_2011_list)
print("Min percentage for 2011 is", min_energy_percentage_2011)

min_energy_percentage_2012 = min(year_2012_list)
print("Min percentage for 2012 is", min_energy_percentage_2012)

#find the corresponding country names
min_index = year_2011_list.index(min_energy_percentage_2011)
print("Country with the minimum percentage in 2011 is", country_list[min_index])

min_index = year_2012_list.index(min_energy_percentage_2012)
print("Country with the minimum percentage in 2012 is", country_list[min_index])



#finding the averages of both 2011 and 2012
count = len(year_2011_list)
total = sum(year_2011_list)
average = total/count
print("Average nuclear and atl. energy use per country for 2011 is", average,"%")

count = len(year_2012_list)
total = sum(year_2012_list)
average = total/count
print("Average nuclear and atl. energy use per country for 2012 is", average,"%")

Max percentage for 2011 is 89.72528057
Max percentage for 2012 is 89.638477
Country with the maximum percentage in 2011 is Iceland
Country with the maximum percentage in 2012 is Iceland
Min percentage for 2011 is 0.0
Min percentage for 2012 is 0.0
Country with the minimum percentage in 2011 is United Arab Emirates
Country with the minimum percentage in 2012 is United Arab Emirates
Average nuclear and atl. energy use per country for 2011 is 8.717280642916696 %
Average nuclear and atl. energy use per country for 2012 is 8.789302387254674 %


The average percentages of nuclear and alternative energy use for 2011 and 2012 were also found.

## Analysis and Conclusion

It was found that the maximum percenatges for 2011 and 2012 were 89.73% and 89.64% (rounded); both of these figures were out of Iceland. The minimum percentages were both 0.0, and one of the countries that produced no alternative or nuclear energy was the UAE. The average nuclear and alternative energy percentage for 2011 and 2012 was 8.72% and 8.79%, respectively. One area for further exploration are the factors that go into how much a nation invests in nuclear and alternative energy. If a country has strong oil reserves, it might not seem beneficial to use other sources of energy. Additonally, some countries are well positioned to use these other sources. For example Iceland, with the high nuclear and alternative energy percentages at over 89%, has a population of only a few hundred thousand and about 90% of its households use geothermal energy according to the UN (https://unchronicle.un.org/article/iceland-s-sustainable-energy-story-model-world); it would be interesting to see this data broken down by the specific source of energy. The average of just under nine percent is interesting, but becuase the amount of energy produced is not taken into account, the two figures do not completely address questions of widespread usage of nuclear and alternative energy sources. This is becuase of the nature of climate change. If a quarter of a superpower's energy comes from nuclear or alternative sources but the other 75% still comes from fossil fuels, more environmental damage is done than a country that has a percentage of ten but produces so much less energy it doesn't matter. On the other hand, this data indicates how much is being invested into cleaner sources of energy, and that creates immidiate posotive effects in terms of climate change. One more area for further exploration are the factors that contribute to the slight increases and decreases (not to mention the dramatic ones) from year to year. For example the average percentage increased slightly from 2011 to 2012. What event or events caused this?

## Aknowledgements
Ms. Sconyers provided me with all of the code present in both data preparation and modeling. A thanks to the International Energy Agency, the Organization for Economic Cooperation and Development, and the Wolrd Bank for collecting and publishing this data (this data was published by the World Bank but it is owned by the IEA and OECD). Also, the United Nations provided information on Iceland's energy production that assisted in analysis.     